File: class_time_table/src/main/java/xyz/jasenon/lab/class_time_table/t_io/handler/SmartBoardTioHandler.java
package xyz.jasenon.lab.class_time_table.t_io.handler;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.tio.core.ChannelContext;
import org.tio.core.TioConfig;
import org.tio.core.exception.TioDecodeException;
import org.tio.core.intf.Packet;
import org.tio.server.intf.TioServerHandler;
import xyz.jasenon.lab.class_time_table.t_io.adapter.TioPacketAdapter;
import xyz.jasenon.lab.class_time_table.t_io.adapter.TioQosAdapter;
import xyz.jasenon.lab.class_time_table.t_io.codec.TioProtocolCodec;
import xyz.jasenon.lab.tioprotocol.PacketHeader;
import xyz.jasenon.lab.tioprotocol.ProtocolPacket;

import java.nio.ByteBuffer;

/**
 * 智能班牌 t-io 处理器
 * 使用 tio-protocol 进行编解码，集成 QoS 机制
 * 
 * @author Jasenon_ce
 */
@Component
public class SmartBoardTioHandler implements TioServerHandler {
    
    private static final Logger LOG = LoggerFactory.getLogger(SmartBoardTioHandler.class);
    
    private final TioProtocolCodec codec;
    private final TioQosAdapter qosAdapter;
    
    public SmartBoardTioHandler(TioProtocolCodec codec, TioQosAdapter qosAdapter) {
        this.codec = codec;
        this.qosAdapter = qosAdapter;
    }
    
    @Override
    public Packet decode(ByteBuffer buffer, int limit, int position, int readableLength, 
                         ChannelContext channelContext) throws TioDecodeException {
        // 检查是否有足够的数据读取头部
        if (readableLength < PacketHeader.HEADER_LENGTH) {
            return null;
        }
        
        // 设置位置
        buffer.position(position);
        
        // 读取魔数（不移动位置）
        int magic = buffer.getInt(position);
        if (magic != PacketHeader.MAGIC_NUMBER) {
            throw new TioDecodeException("魔数不匹配: 0x" + Integer.toHexString(magic));
        }
        
        // 读取长度字段（在头部最后4字节）
        int bodyLength = buffer.getInt(position + 12);
        if (bodyLength < 0) {
            throw new TioDecodeException("消息体长度无效: " + bodyLength);
        }
        
        // 检查是否有完整的数据包
        int totalLength = PacketHeader.HEADER_LENGTH + bodyLength;
        if (readableLength < totalLength) {
            return null;
        }
        
        // 解码数据包
        try {
            TioPacketAdapter packet = codec.decode(buffer);
            if (packet == null) {
                return null;
            }
            
            LOG.debug("收到消息: channel={}, cmdType={}, seqId={}, qos={}, flags={}",
                    channelContext.getClientNode(), packet.getCmdType(), packet.getSeqId(),
                    packet.getQos(), packet.getFlags());
            
            return packet;
        } catch (TioProtocolCodec.DecodeException e) {
            throw new TioDecodeException("解码失败: " + e.getMessage());
        }
    }
    
    @Override
    public ByteBuffer encode(Packet packet, TioConfig tioConfig, ChannelContext channelContext) {
        TioPacketAdapter adapter = (TioPacketAdapter) packet;
        return codec.encode(adapter);
    }
    
    @Override
    public void handler(Packet packet, ChannelContext channelContext) throws Exception {
        TioPacketAdapter adapter = (TioPacketAdapter) packet;
        ProtocolPacket protocolPacket = adapter.getProtocolPacket();
        byte cmdType = protocolPacket.getCmdType();
        
        // 处理 QoS（发送 ACK 如果需要）
        qosAdapter.handleReceived(channelContext, adapter);
        
        // TODO: 业务处理入口
        // 这里将交给业务层处理，当前版本仅记录日志
        LOG.info("收到业务消息: channel={}, cmdType={}, seqId={}", 
                channelContext.getClientNode(), cmdType, protocolPacket.getSeqId());
        
        // 示例：如果是心跳包，直接回复（实际业务中可能不需要）
        if (cmdType == xyz.jasenon.lab.tioprotocol.CommandType.HEARTBEAT) {
            ProtocolPacket heartbeatAck = new ProtocolPacket();
            heartbeatAck.setMagic(PacketHeader.MAGIC_NUMBER);
            heartbeatAck.setVersion(PacketHeader.VERSION);
            heartbeatAck.setCmdType(xyz.jasenon.lab.tioprotocol.CommandType.HEARTBEAT_ACK);
            heartbeatAck.setSeqId(protocolPacket.getSeqId());
            heartbeatAck.setPayload(new byte[0]);
            heartbeatAck.calculateCheckSum();
            qosAdapter.send(channelContext, heartbeatAck);
        }
    }
}

File: class-time-table-client/app/src/main/java/xyz/jasenon/classtimetable/network/SmartBoardTioHandler.kt
package xyz.jasenon.classtimetable.network

import android.util.Log
import org.tio.client.intf.TioClientHandler
import org.tio.core.ChannelContext
import org.tio.core.TioConfig
import org.tio.core.intf.Packet
import xyz.jasenon.classtimetable.network.handler.DoorStatusPacketHandler
import xyz.jasenon.classtimetable.network.handler.PacketHandlerRegistry
import xyz.jasenon.classtimetable.network.handler.TimetablePacketHandler
import xyz.jasenon.classtimetable.protocol.CommandType
import xyz.jasenon.classtimetable.protocol.PacketFlags
import xyz.jasenon.classtimetable.protocol.QosLevel
import xyz.jasenon.classtimetable.protocol.SeqIdGenerator
import xyz.jasenon.classtimetable.protocol.SmartBoardPacket
import java.nio.ByteBuffer

class SmartBoardTioHandler : TioClientHandler {

    /**
     * 编码：将 SmartBoardPacket 转换为 ByteBuffer
     */
    override fun encode(packet: Packet,tioConfig: TioConfig, context: ChannelContext): ByteBuffer {
        val smartPacket = packet as SmartBoardPacket
        val body = smartPacket.payload ?: ByteArray(0)
        val bodyLength = body.size

        // 创建一个ByteBuffer，大小为头部长度 + 消息体长度
        val buffer = ByteBuffer.allocate(SmartBoardPacket.HEADER_LENGTH + bodyLength)
        buffer.order(tioConfig.byteOrder)

        // 写入头部（可空字段用默认值，避免心跳包等未设全头部时 NPE）
        buffer.putInt(smartPacket.magic ?: SmartBoardPacket.MAGIC_NUMBER)
        buffer.put(smartPacket.version ?: 0x01.toByte())
        buffer.put(smartPacket.cmdType ?: 0.toByte())
        buffer.putShort(smartPacket.seqId ?: 0)
        buffer.put(smartPacket.qos ?: QosLevel.AT_MOST_ONCE.value)
        buffer.put(smartPacket.flags ?: PacketFlags.NONE)
        buffer.put(smartPacket.reserved ?: 0.toByte())
        buffer.put(smartPacket.checkSum ?: 0.toByte())
        buffer.putInt(bodyLength) // 写入真实的消息体长度

        // 写入消息体
        if (bodyLength > 0) {
            buffer.put(body)
        }

        buffer.flip()
        return buffer
    }

    /**
     * 解码：从 ByteBuffer 解析出 SmartBoardPacket
     */
    override fun decode(buffer: ByteBuffer, limit: Int, position: Int, readableLength: Int, context: ChannelContext): SmartBoardPacket? {
        if (readableLength < SmartBoardPacket.HEADER_LENGTH) {
            return null // 数据包长度不足，等待更多数据
        }

        val magic = buffer.getInt(position)
        if (magic != SmartBoardPacket.MAGIC_NUMBER) {
            throw Exception("Invalid magic number")
        }

        // 读取头部
        buffer.position(position)
        val header = ByteArray(SmartBoardPacket.HEADER_LENGTH)
        buffer.get(header)

        val resultPacket = SmartBoardPacket()
        val headerBuffer = ByteBuffer.wrap(header).order(buffer.order())
        resultPacket.magic = headerBuffer.getInt()
        resultPacket.version = headerBuffer.get()
        resultPacket.cmdType = headerBuffer.get()
        resultPacket.seqId = headerBuffer.getShort()
        resultPacket.qos = headerBuffer.get()
        resultPacket.flags = headerBuffer.get()
        resultPacket.reserved = headerBuffer.get()
        resultPacket.checkSum = headerBuffer.get()
        val bodyLength = headerBuffer.getInt()

        // 检查消息体是否完整
        if (readableLength < SmartBoardPacket.HEADER_LENGTH + bodyLength) {
            return null // 消息体不完整，等待更多数据
        }

        // 读取消息体
        if (bodyLength > 0) {
            val body = ByteArray(bodyLength)
            buffer.get(body)
            resultPacket.payload = body
        }

        return resultPacket
    }

    init {
        PacketHandlerRegistry.register(CommandType.TIMETABLE_RESP, TimetablePacketHandler())
        PacketHandlerRegistry.register(CommandType.DOOR_STATUS, DoorStatusPacketHandler())
    }

    /**
     * 消息处理：按 cmdType 分发到对应 Handler，Handler 通过 RemoteDataObservable 通知观察者更新
     */
    override fun handler(packet: Packet, context: ChannelContext) {
        val smartPacket = packet as SmartBoardPacket
        val handled = PacketHandlerRegistry.dispatch(smartPacket)
        if (!handled) {
            Log.d("SmartBoardTioHandler", "No handler for cmdType=${smartPacket.cmdType}")
        }
    }

    override fun heartbeatPacket(context: ChannelContext): Packet {
        Log.d("SmartBoardTioHandler", "发送心跳包")
        val heartbeatPacket = SmartBoardPacket()
        heartbeatPacket.setCmdType(CommandType.HEARTBEAT)
        heartbeatPacket.setVersion(0x01.toByte())
        heartbeatPacket.setSeqId(SeqIdGenerator().nextSeqId())
        heartbeatPacket.setQos(QosLevel.AT_MOST_ONCE.value)
        heartbeatPacket.setFlags(PacketFlags.NONE)
        heartbeatPacket.setReserved(0.toByte())
        heartbeatPacket.setPayload(null)
        heartbeatPacket.setLength(0)
        heartbeatPacket.calculateCheckSum()
        return heartbeatPacket
    }
}